bigquery_etl/publish_json.py

```diff
@@ -0,0 +1,148 @@
+            sys.exit(1)
+
+    def __exit__(self):
+        """Delete temporary tables."""
+        if self.temp_table:
+            self.client.delete_table(self.temp_table)
+
+    def publish_json(self):
+        """Publish query results as JSON to GCP Storage bucket."""
+        if self.metadata.is_incremental():
+            if self.date is None:
+                print("Cannot publish JSON. submission_date missing in parameter.")
+                sys.exit(1)
+
+            # if it is an incremental query, then the query result needs to be
+            # written to a temporary table to get exported as JSON
+            self._write_results_to_temp_table()
+            self._publish_table_as_json(self.temp_table)
+        else:
+            # for non-incremental queries, the entire destination table is exported
+            result_table = f"{self.dataset}.{self.table}_{self.version}"
+            self._publish_table_as_json(result_table)
+
+    def _publish_table_as_json(self, result_table):
+        """Export the `result_table` data as JSON to Cloud Storage."""
+        prefix = (
+            f"api/{self.api_version}/tables/{self.dataset}/"
+            f"{self.table}/{self.version}/files/"
+        )
+
+        if self.date is not None:
+            # if date exists, then query is incremental and newest results are exported
+            prefix += f"{self.date}/"
+
+        table_ref = self.client.get_table(result_table)
+
+        job_config = bigquery.ExtractJobConfig()
+        job_config.destination_format = "NEWLINE_DELIMITED_JSON"
+
+        # "*" makes sure that files larger than 1GB get split up into JSON files
+        destination_uri = f"gs://{self.target_bucket}/" + prefix + "*.json"
+        extract_job = self.client.extract_table(
+            table_ref, destination_uri, location="US", job_config=job_config
+        )
+        extract_job.result()
+
+        self._gcp_convert_ndjson_to_json(prefix)
+
+    def _gcp_convert_ndjson_to_json(self, gcp_path):
+        """Convert ndjson files on GCP to json files."""
+        blobs = self.storage_client.list_blobs(self.target_bucket, prefix=gcp_path)
+        bucket = self.storage_client.bucket(self.target_bucket)
+
+        for blob in blobs:
+            blob_path = f"gs://{self.target_bucket}/{blob.name}"
+            tmp_blob = bucket.blob(blob.name + ".tmp")
+            tmp_blob_name = blob_path + ".tmp"
+
+            # stream from GCS
+            with smart_open.open(blob_path) as fin:
+                with smart_open.open(tmp_blob_name, "w") as fout:
+                    fout.write("[\n")
+
+                    first_line = True
+
+                    for line in fin:
+                        if not first_line:
+                            fout.write(",\n")
+
+                        first_line = False
+                        fout.write(line.replace("\n", ""))
+
+                    fout.write("]")
+
+            bucket.rename_blob(tmp_blob, blob.name)
```

Comment:
This may end up a bit complicated but we _should_ gzip these files. Reading through a bit it looks like [smart_open can handle that](https://github.com/RaRe-Technologies/smart_open#supported-compression-formats), but we may need to [set the `Content-Encoding`](https://cloud.google.com/storage/docs/transcoding#content-type_vs_content-encoding).